pkg/controller/controller.go

```diff
@@ -0,0 +1,161 @@
+package controller
+
+import (
+	"github.com/pkg/errors"
+	"github.com/weaveworks/weave-npc/pkg/util/ipset"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	"k8s.io/kubernetes/pkg/util/iptables"
+	"sync"
+)
+
+type NetworkPolicyController interface {
+	AddNamespace(ns *api.Namespace) error
+	UpdateNamespace(oldObj, newObj *api.Namespace) error
+	DeleteNamespace(ns *api.Namespace) error
+
+	AddPod(obj *api.Pod) error
+	UpdatePod(oldObj, newObj *api.Pod) error
+	DeletePod(obj *api.Pod) error
+
+	AddNetworkPolicy(obj *extensions.NetworkPolicy) error
+	UpdateNetworkPolicy(oldObj, newObj *extensions.NetworkPolicy) error
+	DeleteNetworkPolicy(obj *extensions.NetworkPolicy) error
+}
+
+type controller struct {
+	sync.Mutex
+
+	ipt iptables.Interface
+	ips ipset.Interface
+
+	nss         map[string]*ns // ns name -> ns struct
+	nsSelectors *selectorSet   // selector string -> nsSelector
+}
+
+func New(ipt iptables.Interface, ips ipset.Interface) NetworkPolicyController {
+	c := &controller{
+		ipt: ipt,
+		ips: ips,
+		nss: make(map[string]*ns)}
+
+	c.nsSelectors = newSelectorSet(ips, c.onNewNsSelector)
+
+	return c
+}
+
+func (npc *controller) onNewNsSelector(selector *selector) error {
+	for _, ns := range npc.nss {
+		if ns.namespace != nil {
```

Comment:
There are no ordering guarantees regarding when we hear about new Namespace objects vs say Pods that inhabit that namespace - we could have created the `ns` struct because of a pod creation, and then only fill in the k8s Namespace object later on.